controller/controller.go

```diff
@@ -0,0 +1,171 @@
+		w.WriteHeader(http.StatusNotFound)
+		encoderErr := json.NewEncoder(w).Encode(map[string]string{"message": "Pokemon id  not found , Valid id`s 1-801"})
+		if encoderErr != nil {
+			log.Println(encoderErr)
+		}
+	} else {
+		encoderErr := json.NewEncoder(w).Encode(*pokemon)
+		if encoderErr != nil {
+			log.Println(encoderErr)
+		}
+	}
+}
+
+func (c *PokemonController) PostPokemon(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	name := vars["name"]
+
+	if reflect.TypeOf(name).Kind() != reflect.String {
+		w.WriteHeader(http.StatusBadRequest)
+		encoderErr := json.NewEncoder(w).Encode(map[string]string{"error": "Name should be an string"})
+		if encoderErr != nil {
+			log.Println(encoderErr)
+		}
+	}
+
+	message, err := c.usecase.SavePokemonByNameAPI(name)
+
+	if err != nil {
+		w.WriteHeader(http.StatusNotFound)
+		encoderErr := json.NewEncoder(w).Encode(map[string]string{"message": "Pokemon not found"})
+		if encoderErr != nil {
+			log.Println(encoderErr)
+		}
+	} else {
+		encoderErr := json.NewEncoder(w).Encode(map[string]string{"message": message})
+		if encoderErr != nil {
+			log.Println(encoderErr)
+		}
+	}
+}
+
+func (c *PokemonController) GetPokemonsConcurrently(w http.ResponseWriter, r *http.Request) {
+	types := r.URL.Query().Get("type")
+	items := r.URL.Query().Get("items")
+	itemsPW := r.URL.Query().Get("items_per_workers")
+	defer r.Body.Close()
+
+	itemsInt, errItems := strconv.Atoi(items)
+	items_per_worker, errWorker := strconv.Atoi(itemsPW)
+
+	if errItems != nil || errWorker != nil || items_per_worker == 0 || itemsInt == 0 || itemsInt%items_per_worker != 0 || types != "even" && types != "odd" || itemsInt < items_per_worker {
+
+		err := ""
+
+		if types != "even" && types != "odd" {
+			err = "Type should be even or odd"
+
+		} else if errWorker != nil || items_per_worker == 0 {
+			err = "Items_per_workers should be a valid number more than 0"
+
+		} else if errItems != nil || items_per_worker == 0 {
+
+			err = "Items should be a valid number more than 0"
+
+		} else if itemsInt%items_per_worker != 0 {
+			err = "items_per_workers should be a multiple of items"
+		} else if itemsInt < items_per_worker {
+			err = "items should be greather than items_per_workers"
+		}
+
+		w.Header().Add("Content-Type", "application/json")
+		w.WriteHeader(http.StatusBadRequest)
+		encoderErr := json.NewEncoder(w).Encode(map[string]string{"error": err})
+		if encoderErr != nil {
+			log.Println(encoderErr)
+		}
+
+	} else {
+		pokemons, err := c.usecase.GetPokemonsConcurrently(types, itemsInt, items_per_worker)
+
+		if err != nil {
+			w.WriteHeader(http.StatusNotFound)
+			encoderErr := json.NewEncoder(w).Encode(map[string]string{"message": err.Error()})
+
+			if encoderErr != nil {
+				log.Println(encoderErr)
+			}
+		} else {
+			w.Header().Add("Content-Type", "application/json")
+			encodeErr := json.NewEncoder(w).Encode(*pokemons)
+			if encodeErr != nil {
+				log.Println(encodeErr)
+			}
+		}
+	}
+
+}
```

Comment:
This looks complicated and hard to maintain; no need to make all these checks all at once. This makes the code not readable at all. Just imagine there are 2 more query params added to the requirements.

Look at this example, not sure if works or not, but I give you an example on how to check these values, how to error wrap, and one and another simplification.
```suggestion
func (c *PokemonController) GetPokemonsConcurrently(w http.ResponseWriter, r *http.Request) {
	w.Header().Add("Content-Type", "application/json")
	w.WriteHeader(http.StatusBadRequest)

	items, err := strconv.Atoi(r.URL.Query().Get("items"))
	if err != nil {
		err = errors.Wrap(err, "Error converting items param to int")
		if errEncoding := json.NewEncoder(w).Encode(map[string]string{"error": "items should be an integer"}); errEncoding != nil {
			err = errors.Wrap(err, errEncoding.Error())
		}

		log.Println(err)

		return
	}

	itemsPerWorker, err := strconv.Atoi(r.URL.Query().Get("items_per_worker"))
	if err != nil {
		err = errors.Wrap(err, "Error converting items_per_worker param to int")
		if errEncoding := json.NewEncoder(w).Encode(map[string]string{"error": "itemsPerWorker should be an integer"}); errEncoding != nil {
			err = errors.Wrap(err, errEncoding.Error())
		}

		log.Println(err)

		return
	}

	filterType := r.URL.Query().Get("type")

	switch filterType {
	case "odd", "even":
		break
	case "":
		err = errors.New("type param should be non empty")
	default:
		err = errors.New("type param should be either \"odd\" or \"even\"")
	}

	if err != nil {
		err = errors.Wrap(err, "Error getting type param")
		if errEncoding := json.NewEncoder(w).Encode(map[string]string{"error": err.Error()}); errEncoding != nil {
			err = errors.Wrap(err, errEncoding.Error())
		}

		log.Println(err)

		return
	}

	pokemons, err := c.usecase.GetPokemonsConcurrently(filterType, items, itemsPerWorker)
	if err != nil {
		err = errors.Wrap(err, "Error getting pokemons concurrently")
		if errEncoding := json.NewEncoder(w).Encode(map[string]string{"error": err.Error()}); errEncoding != nil {
			err = errors.Wrap(err, errEncoding.Error())
		}

		log.Println(err)

		return
	}

	if pokemons == nil {
		w.WriteHeader(http.StatusNotFound)
		err := errors.New("No pokemons found")
		if errEncoding := json.NewEncoder(w).Encode(map[string]string{"message": err.Error()}); errEncoding != nil {
			err = errors.Wrap(err, errEncoding.Error())
		}

		log.Println(err)

		return
	}

	errEncode := json.NewEncoder(w).Encode(*pokemons)
	if errEncode != nil {
		log.Println(errEncode)
	}
}
```